repos/exporter/lib/atcoCheckerHandler.ts

```diff
@@ -0,0 +1,286 @@
+
+        const pointToPointTicket = JSON.parse(object.Body.toString('utf-8')) as
+            | WithIds<SingleTicket>
+            | WithIds<ReturnTicket>
+            | WithIds<PointToPointPeriodTicket>;
+
+        // check to see if single ticket
+        if ('fareZones' in pointToPointTicket) {
+            const idsOfServicesRequiringAttention: string[] = await getIdsOfServicesRequiringAttentionForSingles(
+                pointToPointTicket,
+            );
+
+            await saveIdsOfServicesRequiringAttentionInTheDb(productId, idsOfServicesRequiringAttention);
+        } else if ('inboundFareZones' in pointToPointTicket && 'outboundFareZones' in pointToPointTicket) {
+            // we have a return ticket
+            const idsOfServicesRequiringAttention: string[] = await getIdsOfServicesRequiringAttentionForReturns(
+                pointToPointTicket,
+            );
+
+            await saveIdsOfServicesRequiringAttentionInTheDb(productId, idsOfServicesRequiringAttention);
+        }
+    }
+};
+
+/**
+ * For single tickets; gets a list of ids for the serices that require
+ * attention because of a change on the stops relating to the service.
+ *
+ * @param {pointToPointTicket} the matching json for single ticket.
+ *
+ * @returns {Promise<string[]>} a string array containing the ids of the
+ * services that require attention.
+ */
+const getIdsOfServicesRequiringAttentionForSingles = async (
+    pointToPointTicket: WithIds<SingleTicket>,
+): Promise<string[]> => {
+    // all stops from the matching JSON
+    let atcoCodesOfKnownStops: string[] = [];
+
+    pointToPointTicket.fareZones.forEach((fareZone) => {
+        fareZone.stops.forEach((stop) => {
+            atcoCodesOfKnownStops.push(stop.atcoCode);
+        });
+    });
+
+    if (pointToPointTicket.unassignedStops.singleUnassignedStops) {
+        pointToPointTicket.unassignedStops.singleUnassignedStops.forEach((stop) => {
+            atcoCodesOfKnownStops.push(stop.atcoCode);
+        });
+    }
+
+    // by this point, atcoCodesOfKnownStops is populated.
+
+    const { journeyDirection, lineId, nocCode } = pointToPointTicket;
+
+    const directionsAndStops = await getDirectionAndStopsByLineIdAndNoc(lineId, nocCode);
+
+    const theStopsMatchingTheJourneyDirectionOnOurTicket = directionsAndStops.filter(
+        (directionsAndStopsItem) => directionsAndStopsItem.direction === journeyDirection,
+    );
+
+    // split by service id and build a hashmap (dictionary) where the key is
+    // the service id and the value is the array of stops
+    const serviceAndStopsFromFeed = buildAHashmapSplitOnServiceId(theStopsMatchingTheJourneyDirectionOnOurTicket);
+
+    // sort the stops and remove duplicates for each service
+    const serviceAndSortedUniqueStopsFromFeed = sortAndRemoveDuplicateStops(serviceAndStopsFromFeed);
+
+    // on our matching json stops, sort the stops and remove duplicates
+    atcoCodesOfKnownStops = [...new Set(atcoCodesOfKnownStops.sort())];
```

Comment:
Although this may work in practice, technically the "order" of a set is not well defined so you are better off sorting the array rather than the Set

Reply:
If interested you can find out more here: http://jsclass.jcoglan.com/set.html

Reply:
I'm confused. **I am** sorting the array? I'm sorting it first, then passing it to the `Set()`?